Added OpenTelemetry related blog post (#343)
strimzi/strimzi.github.io: _posts/2023-03-06-opentelemetry.md

---
layout: post
title: "Apache Kafka tracing with Strimzi: from OpenTracing to OpenTelemetry"
date: 2023-03-06
author: paolo_patierno
---

Apache Kafka is one of the most used platforms for events and data streaming to build a distributed system, by providing the backbone for microservices communication.
One of the most complex tasks to deal with such a distributed environment, is to get insights about the flow of the data and then evaluating performance.
This is where distributed tracing comes into the picture, by addressing different issues:

* monitor the flow of events or remote procedure calls between applications.
* profile messages latency in order to help optimizing performance.
* provide an overall picture about how microservices communicate across boundaries.

> Strimzi have been provided tracing support, for Kafka Connect, Kafka Mirror Maker(s) and the HTTP bridge, since the 0.14.0 release leveraging the [OpenTracing](https://opentracing.io/) project. Tracing is not supported for Kafka brokers.

Reviewer: I think `Tracing is not supported for Kafka brokers.` can be very confusing. So I wonder if some more clarification would make sense.

Reviewer: Yes, I think that is better.

But since then, the distributed tracing ecosystem has changed, the OpenTracing project was archived and replaced by a new project, [OpenTelemetry](https://opentelemetry.io/).

This blog post will show you, how to use OpenTelemetry within the Strimzi project and what are the differences with OpenTracing which is now deprecated and will be removed in the coming months.

## What is OpenTelemetry?

OpenTelemetry is an open source and vendor neutral project providing tools, APIs, and SDKs used to generate, collect and export telemetry data.
It is part of the Cloud Native Computing Foundation and it is available across several languages.
It is not just about tracing but it has support for metrics and logs as well.

> > The Strimzi project supports with OpenTelemetry is about tracing only. There are no actual plans to support OpenTelemetry metrics and logs at this time.

Reviewer: ```suggestion
> The Strimzi project support of OpenTelemetry is about tracing only. There are no actual plans to support OpenTelemetry metrics and logs at this time.
```

You could also decide to use the usual Jaeger protocol endpoint by setting the corresponding Jaeger exporter in your system.

From an application perspective, the way OpenTelemetry works is similar to OpenTracing.
The tracer is part of your application and it provides the API for creating trace spans and using the configured exporter to send them.
Focusing on Apache Kafka, the application could be a producer or consumer, or one of the components managed by the Strimzi operator, such as Kafka Connect, Kafka Mirror Maker(s) and the HTTP bridge.

As it happens with OpenTracing, even the OpenTelemetry exporter doesn't send all traces but often just part of them by using sampling.
The sampler mechanism can be configured through environment variables.

## How to enable OpenTelemetry

Before adding the support for OpenTelemetry, it was already possible to enable the distributed tracing on the `KafkaConnect`, `KafkaMirrorMaker`, `KafkaMirrorMaker2` and `KafkaBridge` custom resources by setting the `spec.tracing.type` property to `jaeger` to use OpenTracing.

> Since the Strimzi 0.32.0 release, you can set the corresponding `opentelemetry` value for using OpenTelemetry instead.

Reviewer: Doesn't 0.32 still use Jaeger instead of OLTP? Wouldn't it be better for the sake of the post to just pretend its 0.33+ as you cover only OLTP?

Author: Yeah you are right, I will fix to 0.33.0 there.

```yaml
# ...
spec:
  # ...
  tracing:
    type: opentelemetry
  # ...  
```

With the above setting, the Strimzi operator initializes the OpenTelemetry tracer using the default OTLP exporter, for sending traces to the tracing backend.
But you also need to specify environment variables to configure the exporter itself, by using the `spec.template` properties.
Following, an example of the `KafkaBridge` configuration.

```yaml
spec:
  #...
  template:
    bridgeContainer:
      env:
        - name: OTEL_SERVICE_NAME
          value: my-otel-service
        - name: OTEL_EXPORTER_OTLP_ENDPOINT
          value: "http://jaeger-host:4317"
  tracing:
    type: opentelemetry
```

It is using the service name `my-otel-service` which will show up in the backend UI.

> The service name is specific to the operand you are using in order to group multiple traces together. You should use a different service name for any other operand (i.e. Kafka Connect, Kafka Mirror Maker, ...)

> Assuming, for example, you are using Jaeger backend, it configures the tracer to send the traces to the `http://jaeger-host:4317` OTLP endpoint on the backend.

Reviewer: Should it explaian what `jaeger-host` is? Or say that its a placeholder for the Jaeger URL?

By getting the [deployment-tracing-opentelemetry.yaml](https://raw.githubusercontent.com/strimzi/client-examples/main/java/http/deployment-tracing-opentelemetry.yaml) YAML file you are able to deploy:

* a producer Java HTTP client which sends a message every second to the Apache Kafka cluster through the Strimzi HTTP bridge.
* a consumer Java HTTP client which polls for messages every second from the Apache Kafka cluster through the Strimzi HTTP bridge.

Both applications are configured in order to trace the HTTP requests they send to the bridge.

Get the above file and change the OpenTelemetry exporter configuration environment variables to match with what we have already deployed and configured in the `KafkaBridge` custom resource.

```yaml
env:
  # ...
  - name: OTEL_SERVICE_NAME
    value: my-otel-service
  - name: OTEL_EXPORTER_OTLP_ENDPOINT
    value: http://simplest-collector:4317
  # ...
```

Save and deploy the YAML:

```shell
kubectl apply -f deployment-tracing-opentelemetry.yaml
```

When the HTTP clients exchange the messages we can see the traces and the corresponding spans in the Jaeger Web UI.

> ![](/assets/images/posts/2023-03-06-opentelemetry-01.png)

Reviewer: Shouldn't the clients be visible in the traces as well?

Author: The clients sending and receiving are part of the above two traces. If you see at the other pictures, you can see both the producer and consumer traces have:

* one span about the HTTP client sending the request
* one span about the bridge handling the HTTP request
* one span about the Kafka client send/receive to/from the topic

Reviewer: Not sure I follow. It is 3 spans. But all of them are from the same `my-otel-service`, so they all come from the bridge. The producer / consumer is not visible there.

Author: I set the same `my-otel-service` for the HTTP clients applications as well.
They are producing that spans under the same trace. From a UI perspective you don't have to jump across services in the drop down list.

Reviewer: That seems a bit strange and the flow is not obvious from it. But I guess if that is how you want it, then fine. 🤷

Author: I got what you mean, going to run with different service names and catching a new screenshot.

Author: @scholzj just pushed changes, let me know what do you think now please.
